### PR TITLE
Enable using different Tor identities per node.

### PR DESCRIPTION
**Pull request by nopara73**
Closes https://github.com/zkSNACKs/WalletWasabi/issues/1293

**UPDATE:** This PR is tested and ready.

The code is probably good, but I couldn't test it. Is there a quick way to test it or should I tie together my local NBitcoin branch with Wasabi somehow?  

If it doesn't fail, then it works as intended since the code has been brought from Wasabi where we have tests for it.  

The trick is that Tor SOCKS5 authentication is actually not authentication, but you can give it any string and if the string is different, then Tor will use a different Tor stream. (With best effort.)

### CODE REVIEW

**Comment by nopara73** [comment]
@lontivero Could you test if this code works somehow? I think you have already tied together Wasabi with a local NBitcoin branch of yours.

File: NBitcoin/Socks/SocksHelper.cs [hunk 1]
```diff
@@ -66,17 +76,57 @@ internal static byte[] CreateConnectMessage(EndPoint endpoint)
...
 			await stream.FlushAsync().WithCancellation(cancellationToken).ConfigureAwait(false);
 
 			var selectionResponse = new byte[2];
 			await stream.ReadAsync(selectionResponse, 0, 2).WithCancellation(cancellationToken).ConfigureAwait(false);
 			if (selectionResponse[0] != 5)
 				throw new SocksException("Invalid version in selection reply");
-			if (selectionResponse[1] != 0)
+			if (selectionResponse[1] == 2)
+			{
+				// https://tools.ietf.org/html/rfc1929#section-2
```
Comment (nopara73) [inline NBitcoin/Socks/SocksHelper.cs]: Basically testing should be just putting a breakpoint here and if it gets this far and it doesn't throw any exceptions in the rest of the `Handshake` function then the code works as intended.

**Comment by lontivero** [comment]
@nopara73 I've tested with a disabled test already existing in nbitcoin and it and it works. I don't know how to update your PR. My changed UT is in this commit https://github.com/lontivero/NBitcoin/commit/bcbc29b721da8eadccf94df7c55df9a1bab7a988

**Comment by nopara73** [comment]
@lontivero Here, I create a PR from your patch and merged it: https://github.com/nopara73/NBitcoin/pull/1

File: NBitcoin/Socks/SocksHelper.cs [hunk 2]
```diff
@@ -66,17 +76,57 @@ internal static byte[] CreateConnectMessage(EndPoint endpoint)
...
+
+				// Create a random string 21 char string.
+				const string chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
+				var identity = new string(Enumerable.Repeat(chars, 21)
+				  .Select(s => s[new Random().Next(s.Length)]).ToArray());
+
+				var username = identity;
+				var password = identity;
+				var uName = Encoding.UTF8.GetBytes(username);
+				var passwd = Encoding.UTF8.GetBytes(password);
```
Comment (NicolasDorier) [inline NBitcoin/Socks/SocksHelper.cs]: are you sure about this? UTF8 by default insert a BOM marker as first byte.

File: NBitcoin/Socks/SocksHelper.cs [hunk 3]
```diff
@@ -66,17 +76,57 @@ internal static byte[] CreateConnectMessage(EndPoint endpoint)
...
+				// https://tools.ietf.org/html/rfc1929#section-2
+				// Once the SOCKS V5 server has started, and the client has selected the
+				// Username / Password Authentication protocol, the Username / Password
+				// subnegotiation begins.  This begins with the client producing a
+				// Username / Password request:
+
+				// Create a random string 21 char string.
+				const string chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
+				var identity = new string(Enumerable.Repeat(chars, 21)
+				  .Select(s => s[new Random().Next(s.Length)]).ToArray());
```
Comment (NicolasDorier) [inline NBitcoin/Socks/SocksHelper.cs]: If you run this, you will get `AAAAAABBB8888888888QQQQQ` kind of password, so you need to create `new Random()` only once.

File: NBitcoin/Socks/SocksHelper.cs [hunk 4]
```diff
@@ -66,17 +76,57 @@ internal static byte[] CreateConnectMessage(EndPoint endpoint)
 			return sendBuffer;
 		}
 
-		public static async Task Handshake(Socket socket, EndPoint endpoint, CancellationToken cancellationToken)
+		public static async Task Handshake(Socket socket, EndPoint endpoint, bool changeTorIdentities, CancellationToken cancellationToken)
```
Comment (NicolasDorier) [inline NBitcoin/Socks/SocksHelper.cs]: Can you ask the user to pass NetworkCredentials here ?

File: NBitcoin/Socks/SocksHelper.cs (hunk 3 above)
Comment (knocte) [reply NBitcoin/Socks/SocksHelper.cs]: why not just create a new `Guid` and truncate it?

**Comment by NicolasDorier** [comment]
I will update this PR myself, there is several issues with it I commented on.

File: NBitcoin/Socks/SocksHelper.cs [hunk 5]
```diff
@@ -66,17 +76,57 @@ internal static byte[] CreateConnectMessage(EndPoint endpoint)
...
+			if (selectionResponse[1] == 2)
+			{
+				// https://tools.ietf.org/html/rfc1929#section-2
+				// Once the SOCKS V5 server has started, and the client has selected the
+				// Username / Password Authentication protocol, the Username / Password
+				// subnegotiation begins.  This begins with the client producing a
+				// Username / Password request:
+
+				// Create a random string 21 char string.
+				const string chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
```
Comment (NicolasDorier) [inline NBitcoin/Socks/SocksHelper.cs]: did not knew you could put const here

File: NBitcoin.Tests/ProtocolTests.cs [hunk 6]
```diff
@@ -839,7 +841,7 @@ public async Task TestDifferentConnectionMethods()
 				// Should works for DNS names
 				"ec2-52-14-64-82.us-east-2.compute.amazonaws.com"
 				};
-			foreach (var onlyForOnionHosts in new[] { true, false })
+			foreach (var (onlyForOnionHosts, changeIpIdentities) in new[] { (true, true), (true, false), (false, true), (false, false) })
```
Comment (NicolasDorier) [inline NBitcoin.Tests/ProtocolTests.cs]: cool stuff, did not knew it was possible

File: NBitcoin/Socks/SocksHelper.cs (hunk 2 above)
Comment (lontivero) [reply NBitcoin/Socks/SocksHelper.cs]: BOM is inserted as first char in a text when you do `Encoding.UTF8.GetString` but here we are getting the bytes.

**Comment by NicolasDorier** [comment]
I updated your PR to my taste, merging it

**Comment by NicolasDorier** [comment]
Pushed .97.

File: NBitcoin/Socks/SocksHelper.cs (hunk 2 above)
Comment (nopara73) [reply NBitcoin/Socks/SocksHelper.cs]: Yes, I'm sure. Both in my `PasswdField` and `UNameField` I have this comment in the code: `// Tor accepts UTF8 encoded passwd`

I have this comment in the code: https://github.com/zkSNACKs/WalletWasabi/blob/master/WalletWasabi/TorSocks5/Models/Fields/ByteArrayFields/PasswdField.cs#L13

**Pull request merged**